### DISCUSSION TITLE
Compare grblHAL to FluidNC

### DISCUSSION THREAD

**Discussion by stevempotter** [grblHAL/core#282] [Q&A]
FluidNC seems to have many of the same goals as this project. It only runs on ESP32s but has some hardware abstraction for other parts of a CNC system.  Is there anyone out there who can mention things grblHAL can do or has, which FluidNC can not or does not?

**Comment by terjeio** [comment]
_"can not"_ is impossible to say anything about - FluidNC can evolve in ways I am not able to predict.

In not particular order grblHAL:

* has a completely hardware agnostic core in a separate project/repo.
* provides an architecture and a programming evironment that facilitates [plugin development](https://github.com/grblHAL/Templates).
* has drivers for many [microprocessors](https://github.com/grblHAL/drivers) and [boards](https://github.com/grblHAL/Controllers), also provided by 3rd parties such as the [STM32H7xx driver](https://github.com/dresco/STM32H7xx).
* has many "standard" [plugin options](https://github.com/grblHAL/plugins).
* supports up to 8 axes.
* supports ABC to UVW remapping.
* supports G59.1 - G59.3 coordinate systems.
* supports lathe specific commands: G7, G8, G33, G76 and G96.
* supports many canned cycles: G73, G81, G82, G83, G85, G86 and G89.
* supports cubic splines: G5 and G5.1.
* has ethernet support.
* has native USB virtual serial port support providing high data transfer rates.
* supports a number of tool change modes.
* has fixes for some obscure legacy Grbl "bugs".
* supports [WebUI](https://github.com/grblHAL/Plugin_WebUI) v2 & v3.
* still uses legacy Grbl $$ commands for settings.
* has a [Web Builder](http://svn.io-engineering.com:8080/) for some of the supported processors.
* has backlash compensation support.
* supports parameters and expressions, with LinuxCNC extensions. Conditionals are not supported. NIST RS274NGC - see sections 3.3.2.2 - 3.3.2.4.

There may be other features that FluidNC does not have, and it may have some that I have listed above.
Note that some of the features listed requires driver support - driver capability varies quite a bit.

grblHAL does not:

* provide a single(?) configurable binary by uploading a config file to the controller. 
* support notification protocols.
* ??

Edit: added backlash compensation & parameter and expression support.
